File: bindings/nim/seatuya.nim
##
## seatuya.nim -- Nim FFI bindings for libseatuya
##
## Loads libseatuya at runtime via Nim's dynlib module.
## Set the SEATUYA_LIB environment variable to override the library path.
##
## Usage:
##   import seatuya
##   let dev = tuyaCreate("dev_id", "192.168.1.100", "local_key", "3.3")
##
## Compile: nim c -r example.nim

import dynlib, os, strutils

# ==================================================================
#  Library loading
# ==================================================================

var libHandle: LibHandle

proc loadLibrary(): bool =
  if libHandle != nil: return true
  let path = getEnv("SEATUYA_LIB",
    when defined(windows): "seatuya.dll"
    elif defined(macosx):  "libseatuya.dylib"
    else:                  "libseatuya.so")
  libHandle = loadLib(path)
  if libHandle == nil:
    stderr.writeLine "Failed to load libseatuya: ", path
    return false
  result = true

proc symAddr(name: cstring): pointer =
  discard loadLibrary()
  result = libHandle.symAddr(name)
  if result == nil:
    stderr.writeLine "Symbol not found: ", name

# ==================================================================
#  Helper: consume malloc'd C string and free it
# ==================================================================

proc consumeCString(p: pointer): string =
  ## Convert a malloc'd C string to a Nim string and free it.
  if p == nil: return nil
  result = $cast[cstring](p)
  let freeFn = cast[proc(s: cstring) {.cdecl.}](libHandle.symAddr("tuya_free_string"))
  if freeFn != nil:
    freeFn(cast[cstring](p))

# ==================================================================
#  C function pointer types and wrappers
# ==================================================================

type
  TuyaDevice = object
  TuyaDevicePtr = ptr TuyaDevice

# -- Version --------------------------------------------------------

proc tuyaVersion*(): string =
  let fn = cast[proc(): cstring {.cdecl.}](symAddr("tuya_version"))
  result = $fn()

# -- Lifecycle ------------------------------------------------------

proc tuyaCreate*(deviceId, address, localKey, version: string): TuyaDevicePtr =
  let fn = cast[proc(a, b, c, d: cstring): TuyaDevicePtr {.cdecl.}](symAddr("tuya_create"))
  result = fn(deviceId.cstring, address.cstring, localKey.cstring, version.cstring)

proc tuyaAlloc*(version: string): TuyaDevicePtr =
  let fn = cast[proc(v: cstring): TuyaDevicePtr {.cdecl.}](symAddr("tuya_alloc"))
  result = fn(version.cstring)

proc tuyaDestroy*(dev: TuyaDevicePtr) =
  let fn = cast[proc(d: TuyaDevicePtr) {.cdecl.}](symAddr("tuya_destroy"))
  fn(dev)

# -- Credentials ----------------------------------------------------

proc tuyaSetCredentials*(dev: TuyaDevicePtr; deviceId, localKey: string) =
  let fn = cast[proc(d: TuyaDevicePtr; a, b: cstring) {.cdecl.}](symAddr("tuya_set_credentials"))
  fn(dev, deviceId.cstring, localKey.cstring)

proc tuyaGetDeviceId*(dev: TuyaDevicePtr): string =
  let fn = cast[proc(d: TuyaDevicePtr): cstring {.cdecl.}](symAddr("tuya_get_device_id"))
  result = $fn(dev)

proc tuyaGetLocalKey*(dev: TuyaDevicePtr): string =
  let fn = cast[proc(d: TuyaDevicePtr): cstring {.cdecl.}](symAddr("tuya_get_local_key"))
  result = $fn(dev)

proc tuyaGetIp*(dev: TuyaDevicePtr): string =
  let fn = cast[proc(d: TuyaDevicePtr): cstring {.cdecl.}](symAddr("tuya_get_ip"))
  result = $fn(dev)

# -- Connection -----------------------------------------------------

proc tuyaConnect*(dev: TuyaDevicePtr; hostname: string): bool =
  let fn = cast[proc(d: TuyaDevicePtr; h: cstring): bool {.cdecl.}](symAddr("tuya_connect"))
  result = fn(dev, hostname.cstring)

proc tuyaDisconnect*(dev: TuyaDevicePtr) =
  let fn = cast[proc(d: TuyaDevicePtr) {.cdecl.}](symAddr("tuya_disconnect"))
  fn(dev)

proc tuyaIsConnected*(dev: TuyaDevicePtr): bool =
  let fn = cast[proc(d: TuyaDevicePtr): bool {.cdecl.}](symAddr("tuya_is_connected"))
  result = fn(dev)

proc tuyaReconnect*(dev: TuyaDevicePtr): bool =
  let fn = cast[proc(d: TuyaDevicePtr): bool {.cdecl.}](symAddr("tuya_reconnect"))
  result = fn(dev)

# -- Retry ----------------------------------------------------------

proc tuyaSetRetryLimit*(dev: TuyaDevicePtr; limit: int32) =
  let fn = cast[proc(d: TuyaDevicePtr; n: int32) {.cdecl.}](symAddr("tuya_set_retry_limit"))
  fn(dev, limit)

proc tuyaSetRetryDelay*(dev: TuyaDevicePtr; delayMs: int32) =
  let fn = cast[proc(d: TuyaDevicePtr; n: int32) {.cdecl.}](symAddr("tuya_set_retry_delay"))
  fn(dev, delayMs)

proc tuyaGetRetryLimit*(dev: TuyaDevicePtr): int32 =
  let fn = cast[proc(d: TuyaDevicePtr): int32 {.cdecl.}](symAddr("tuya_get_retry_limit"))
  result = fn(dev)

proc tuyaGetRetryDelay*(dev: TuyaDevicePtr): int32 =
  let fn = cast[proc(d: TuyaDevicePtr): int32 {.cdecl.}](symAddr("tuya_get_retry_delay"))
  result = fn(dev)

# -- Session negotiation --------------------------------------------

proc tuyaNegotiateSession*(dev: TuyaDevicePtr; localKey: string): bool =
  let fn = cast[proc(d: TuyaDevicePtr; k: cstring): bool {.cdecl.}](symAddr("tuya_negotiate_session"))
  result = fn(dev, localKey.cstring)

proc tuyaNegotiateSessionStart*(dev: TuyaDevicePtr; localKey: string): bool =
  let fn = cast[proc(d: TuyaDevicePtr; k: cstring): bool {.cdecl.}](symAddr("tuya_negotiate_session_start"))
  result = fn(dev, localKey.cstring)

proc tuyaNegotiateSessionFinalize*(dev: TuyaDevicePtr; buf: pointer; size: int32; localKey: string): bool =
  let fn = cast[proc(d: TuyaDevicePtr; b: pointer; s: int32; k: cstring): bool {.cdecl.}](symAddr("tuya_negotiate_session_finalize"))
  result = fn(dev, buf, size, localKey.cstring)

# -- State queries --------------------------------------------------

proc tuyaGetProtocol*(dev: TuyaDevicePtr): int32 =
  let fn = cast[proc(d: TuyaDevicePtr): int32 {.cdecl.}](symAddr("tuya_get_protocol"))
  result = fn(dev)

proc tuyaGetSessionState*(dev: TuyaDevicePtr): int32 =
  let fn = cast[proc(d: TuyaDevicePtr): int32 {.cdecl.}](symAddr("tuya_get_session_state"))
  result = fn(dev)

proc tuyaGetSocketState*(dev: TuyaDevicePtr): int32 =
  let fn = cast[proc(d: TuyaDevicePtr): int32 {.cdecl.}](symAddr("tuya_get_socket_state"))
  result = fn(dev)

proc tuyaGetLastError*(dev: TuyaDevicePtr): int32 =
  let fn = cast[proc(d: TuyaDevicePtr): int32 {.cdecl.}](symAddr("tuya_get_last_error"))
  result = fn(dev)

# -- Async mode -----------------------------------------------------

proc tuyaSetAsyncMode*(dev: TuyaDevicePtr; asyncMode: bool) =
  let fn = cast[proc(d: TuyaDevicePtr; a: bool) {.cdecl.}](symAddr("tuya_set_async_mode"))
  fn(dev, asyncMode)

proc tuyaIsSocketReadable*(dev: TuyaDevicePtr): bool =
  let fn = cast[proc(d: TuyaDevicePtr): bool {.cdecl.}](symAddr("tuya_is_socket_readable"))
  result = fn(dev)

proc tuyaIsSocketWritable*(dev: TuyaDevicePtr): bool =
  let fn = cast[proc(d: TuyaDevicePtr): bool {.cdecl.}](symAddr("tuya_is_socket_writable"))
  result = fn(dev)

proc tuyaSetSessionReady*(dev: TuyaDevicePtr): bool =
  let fn = cast[proc(d: TuyaDevicePtr): bool {.cdecl.}](symAddr("tuya_set_session_ready"))
  result = fn(dev)

# -- Message building/decoding --------------------------------------

proc tuyaBuildMessage*(dev: TuyaDevicePtr; buf: pointer; cmd: int32; payload, key: string): int32 =
  let fn = cast[proc(d: TuyaDevicePtr; b: pointer; c: int32; p, k: cstring): int32 {.cdecl.}](symAddr("tuya_build_message"))
  result = fn(dev, buf, cmd, payload.cstring, key.cstring)

proc tuyaDecodeMessage*(dev: TuyaDevicePtr; buf: pointer; size: int32; key: string): string =
  let fn = cast[proc(d: TuyaDevicePtr; b: pointer; s: int32; k: cstring): pointer {.cdecl.}](symAddr("tuya_decode_message"))
  result = consumeCString(fn(dev, buf, size, key.cstring))

proc tuyaGeneratePayload*(dev: TuyaDevicePtr; cmd: int32; deviceId, datapoints: string): string =
  let fn = cast[proc(d: TuyaDevicePtr; c: int32; did, dps: cstring): pointer {.cdecl.}](symAddr("tuya_generate_payload"))
  result = consumeCString(fn(dev, cmd, deviceId.cstring, datapoints.cstring))

# -- Raw send/receive ------------------------------------------------

proc tuyaSend*(dev: TuyaDevicePtr; buf: pointer; size: int32): int32 =
  let fn = cast[proc(d: TuyaDevicePtr; b: pointer; s: int32): int32 {.cdecl.}](symAddr("tuya_send"))
  result = fn(dev, buf, size)

proc tuyaReceive*(dev: TuyaDevicePtr; buf: pointer; maxsize, minsize: int32): int32 =
  let fn = cast[proc(d: TuyaDevicePtr; b: pointer; mx, mn: int32): int32 {.cdecl.}](symAddr("tuya_receive"))
  result = fn(dev, buf, maxsize, minsize)

# -- device22 mode ---------------------------------------------------

proc tuyaSetDevice22*(dev: TuyaDevicePtr; nullDpsJson: string) =
  let fn = cast[proc(d: TuyaDevicePtr; j: cstring) {.cdecl.}](symAddr("tuya_set_device22"))
  fn(dev, nullDpsJson.cstring)

proc tuyaIsDevice22*(dev: TuyaDevicePtr): bool =
  let fn = cast[proc(d: TuyaDevicePtr): bool {.cdecl.}](symAddr("tuya_is_device22"))
  result = fn(dev)

# -- High-level round-trip -------------------------------------------

proc tuyaSetValueBool*(dev: TuyaDevicePtr; dp: int32; value: bool): string =
  let fn = cast[proc(d: TuyaDevicePtr; dp: int32; v: bool): pointer {.cdecl.}](symAddr("tuya_set_value_bool"))
  result = consumeCString(fn(dev, dp, value))

proc tuyaSetValueInt*(dev: TuyaDevicePtr; dp, value: int32): string =
  let fn = cast[proc(d: TuyaDevicePtr; dp, v: int32): pointer {.cdecl.}](symAddr("tuya_set_value_int"))
  result = consumeCString(fn(dev, dp, value))

proc tuyaSetValueString*(dev: TuyaDevicePtr; dp: int32; value: string): string =
  let fn = cast[proc(d: TuyaDevicePtr; dp: int32; v: cstring): pointer {.cdecl.}](symAddr("tuya_set_value_string"))
  result = consumeCString(fn(dev, dp, value.cstring))

proc tuyaSetValueFloat*(dev: TuyaDevicePtr; dp: int32; value: float64): string =
  let fn = cast[proc(d: TuyaDevicePtr; dp: int32; v: float64): pointer {.cdecl.}](symAddr("tuya_set_value_float"))
  result = consumeCString(fn(dev, dp, value))

proc tuyaTurnOn*(dev: TuyaDevicePtr; switchDp: int32 = 1): string =
  let fn = cast[proc(d: TuyaDevicePtr; dp: int32): pointer {.cdecl.}](symAddr("tuya_turn_on"))
  result = consumeCString(fn(dev, switchDp))

proc tuyaTurnOff*(dev: TuyaDevicePtr; switchDp: int32 = 1): string =
  let fn = cast[proc(d: TuyaDevicePtr; dp: int32): pointer {.cdecl.}](symAddr("tuya_turn_off"))
  result = consumeCString(fn(dev, switchDp))

proc tuyaStatus*(dev: TuyaDevicePtr): string =
  let fn = cast[proc(d: TuyaDevicePtr): pointer {.cdecl.}](symAddr("tuya_status"))
  result = consumeCString(fn(dev))

proc tuyaHeartbeat*(dev: TuyaDevicePtr): string =
  let fn = cast[proc(d: TuyaDevicePtr): pointer {.cdecl.}](symAddr("tuya_heartbeat"))
  result = consumeCString(fn(dev))

# -- Memory ----------------------------------------------------------

proc tuyaFreeString*(s: cstring) =
  let fn = cast[proc(s: cstring) {.cdecl.}](symAddr("tuya_free_string"))
  if fn != nil: fn(s)

# ==================================================================
#  Type-aware set_value dispatcher
# ==================================================================
#
#  tuyaSetValue(dev, dp, "bool",   true)
#  tuyaSetValue(dev, dp, "int",    42)
#  tuyaSetValue(dev, dp, "string", "hello")
#  tuyaSetValue(dev, dp, "float",  3.14)
#

proc tuyaSetValue*[T](dev: TuyaDevicePtr; dp: int32; typ: string; value: T): string =
  case typ
  of "bool":   tuyaSetValueBool(dev, dp, cast[bool](value))
  of "int":    tuyaSetValueInt(dev, dp, cast[int32](value))
  of "string": tuyaSetValueString(dev, dp, cast[string](value))
  of "float":  tuyaSetValueFloat(dev, dp, cast[float64](value))
  else: raise newException(ValueError, "Unknown type: " & typ)

when isMainModule:
  ## Self-test (run with: nim c -r seatuya.nim)
  echo "seatuya.nim loaded. Library path: ", getEnv("SEATUYA_LIB", "libseatuya.so")

# ==================================================================
#  Constants
# ==================================================================

# Protocol versions
const
  TUYA_PROTO_V31* = 0
  TUYA_PROTO_V33* = 1
  TUYA_PROTO_V34* = 2
  TUYA_PROTO_V35* = 3

# Session states
const
  TUYA_SESSION_INVALID*     = 0
  TUYA_SESSION_STARTING*    = 1
  TUYA_SESSION_FINALIZING*  = 2
  TUYA_SESSION_ESTABLISHED* = 3

# Socket states
const
  TUYA_SOCK_NO_SUCH_HOST*  = 0
  TUYA_SOCK_NO_SOCK_AVAIL* = 1
  TUYA_SOCK_FAILED*        = 2
  TUYA_SOCK_DISCONNECTED*  = 3
  TUYA_SOCK_CONNECTING*    = 4
  TUYA_SOCK_CONNECTED*     = 5
  TUYA_SOCK_READY*         = 6
  TUYA_SOCK_RECEIVING*     = 7

# Misc
const
  TUYA_DEFAULT_PORT*        = 6668
  TUYA_RECOMMENDED_BUFSIZE* = 1024
  TUYA_DEFAULT_RETRY_LIMIT* = 5
  TUYA_DEFAULT_RETRY_DELAY* = 100

# Tuya command types (all 45)
const
  TUYA_CMD_UDP*              = 0
  TUYA_CMD_AP_CONFIG*        = 1
  TUYA_CMD_ACTIVE*           = 2
  TUYA_CMD_BIND*             = 3
  TUYA_CMD_RENAME_GW*        = 4
  TUYA_CMD_RENAME_DEVICE*    = 5
  TUYA_CMD_UNBIND*           = 6
  TUYA_CMD_CONTROL*          = 7
  TUYA_CMD_STATUS*           = 8
  TUYA_CMD_HEART_BEAT*       = 9
  TUYA_CMD_DP_QUERY*         = 10
  TUYA_CMD_QUERY_WIFI*       = 11
  TUYA_CMD_TOKEN_BIND*       = 12
  TUYA_CMD_CONTROL_NEW*      = 13
  TUYA_CMD_ENABLE_WIFI*      = 14
  TUYA_CMD_DP_QUERY_NEW*     = 16
  TUYA_CMD_SCENE_EXECUTE*    = 17
  TUYA_CMD_UPDATEDPS*        = 18
  TUYA_CMD_UDP_NEW*          = 19
  TUYA_CMD_AP_CONFIG_NEW*    = 20
  TUYA_CMD_GET_LOCAL_TIME*   = 28
  TUYA_CMD_WEATHER_OPEN*     = 32
  TUYA_CMD_WEATHER_DATA*     = 33
  TUYA_CMD_STATE_UPLOAD_SYN* = 34
  TUYA_CMD_STATE_UPLOAD_SYN_RECV* = 35
  TUYA_CMD_HEART_BEAT_STOP*  = 37
  TUYA_CMD_STREAM_TRANS*     = 38
  TUYA_CMD_GET_WIFI_STATUS*  = 43
  TUYA_CMD_WIFI_CONNECT_TEST* = 44
  TUYA_CMD_GET_MAC*          = 45
  TUYA_CMD_GET_IR_STATUS*    = 46
  TUYA_CMD_IR_TX_RX_TEST*    = 47
  TUYA_CMD_LAN_GW_ACTIVE*    = 240
  TUYA_CMD_LAN_SUB_DEV_REQUEST* = 241
  TUYA_CMD_LAN_DELETE_SUB_DEV*  = 242
  TUYA_CMD_LAN_REPORT_SUB_DEV*  = 243
  TUYA_CMD_LAN_SCENE*        = 244
  TUYA_CMD_LAN_PUBLISH_CLOUD_CONFIG* = 245
  TUYA_CMD_LAN_PUBLISH_APP_CONFIG*   = 246
  TUYA_CMD_LAN_EXPORT_APP_CONFIG*    = 247
  TUYA_CMD_LAN_PUBLISH_SCENE_PANEL*  = 248
  TUYA_CMD_LAN_REMOVE_GW*   = 249
  TUYA_CMD_LAN_CHECK_GW_UPDATE* = 250
  TUYA_CMD_LAN_GW_UPDATE*   = 251
  TUYA_CMD_LAN_SET_GW_CHANNEL* = 252

File: bindings/clojure/seatuya.clj
;; seatuya.clj -- Clojure JNA bindings for libseatuya
;;
;; Usage: (require 'seatuya)
;;
;; Set SEATUYA_LIB environment variable to override the library path.
;; Requires JNA (com.sun.jna) on the classpath.
;;
;; Run: java -cp jna.jar:clojure.jar clojure.main -i seatuya.clj
;; Or from Leiningen:
;;   :dependencies [[net.java.dev.jna/jna "5.14.0"]]

(ns seatuya
  (:import (com.sun.jna NativeLibrary Function Pointer)
           (java.util Map HashMap)))

;; ==================================================================
;;  Library loading with SEATUYA_LIB env var support
;; ==================================================================

(defonce ^:private native-lib
  (delay
    (let [path (or (System/getenv "SEATUYA_LIB")
                   (let [os (-> (System/getProperty "os.name") (.toLowerCase))]
                     (cond (.contains os "win")  "seatuya.dll"
                           (.contains os "mac")  "libseatuya.dylib"
                           :else                 "libseatuya.so")))]
      (try
        (NativeLibrary/getInstance path)
        (catch Exception e
          (throw (RuntimeException.
                   (str "Failed to load libseatuya from " path ": " (.getMessage e)))))))))

(defn- get-fn [name]
  (.getFunction @native-lib name))

(defn- invoke [ret-type name & args]
  (.invoke (get-fn name) ret-type (into-array Object args)))

;; ==================================================================
;;  Internal helpers
;; ==================================================================

(defn- ptr->bool [x]
  "Convert a JNA boolean (or nil) to Clojure boolean."
  (if x true false))

(defn- consume-ptr
  "Convert a Pointer (malloc'd C string) to Clojure string and free it."
  [ptr]
  (when (and ptr (not (.equals ptr Pointer/NULL)))
    (let [s (.getString ptr 0 "UTF-8")]
      (.invoke (get-fn "tuya_free_string") Void (into-array Object [ptr]))
      s)))

;; ==================================================================
;;  Public API
;; ==================================================================

;; -- Version --------------------------------------------------------

(defn tuya-version []
  (invoke String "tuya_version"))

;; -- Lifecycle ------------------------------------------------------

(defn tuya-create
  "Create a device handle, connect, and negotiate session.
  Returns a Pointer to the device, or nil on failure."
  [device-id address local-key version]
  (let [ptr (invoke Pointer "tuya_create" device-id address local-key version)]
    (when (and ptr (not (.equals ptr Pointer/NULL))) ptr)))

(defn tuya-alloc
  "Allocate a device handle without connecting."
  [version]
  (let [ptr (invoke Pointer "tuya_alloc" version)]
    (when (and ptr (not (.equals ptr Pointer/NULL))) ptr)))

(defn tuya-destroy [dev]
  (invoke Void "tuya_destroy" dev))

;; -- Credentials ----------------------------------------------------

(defn tuya-set-credentials [dev device-id local-key]
  (invoke Void "tuya_set_credentials" dev device-id local-key))

(defn tuya-get-device-id [dev]
  (invoke String "tuya_get_device_id" dev))

(defn tuya-get-local-key [dev]
  (invoke String "tuya_get_local_key" dev))

(defn tuya-get-ip [dev]
  (invoke String "tuya_get_ip" dev))

;; -- Connection -----------------------------------------------------

(defn tuya-connect [dev hostname]
  (ptr->bool (invoke Boolean "tuya_connect" dev hostname)))

(defn tuya-disconnect [dev]
  (invoke Void "tuya_disconnect" dev))

(defn tuya-is-connected [dev]
  (ptr->bool (invoke Boolean "tuya_is_connected" dev)))

(defn tuya-reconnect [dev]
  (ptr->bool (invoke Boolean "tuya_reconnect" dev)))

;; -- Retry ----------------------------------------------------------

(defn tuya-set-retry-limit [dev limit]
  (invoke Void "tuya_set_retry_limit" dev (int limit)))

(defn tuya-set-retry-delay [dev delay-ms]
  (invoke Void "tuya_set_retry_delay" dev (int delay-ms)))

(defn tuya-get-retry-limit [dev]
  (invoke Integer "tuya_get_retry_limit" dev))

(defn tuya-get-retry-delay [dev]
  (invoke Integer "tuya_get_retry_delay" dev))

;; -- Session negotiation --------------------------------------------

(defn tuya-negotiate-session [dev local-key]
  (ptr->bool (invoke Boolean "tuya_negotiate_session" dev local-key)))

(defn tuya-negotiate-session-start [dev local-key]
  (ptr->bool (invoke Boolean "tuya_negotiate_session_start" dev local-key)))

(defn tuya-negotiate-session-finalize [dev buf size local-key]
  (ptr->bool (invoke Boolean "tuya_negotiate_session_finalize" dev buf (int size) local-key)))

;; -- State queries --------------------------------------------------

(defn tuya-get-protocol [dev]
  (invoke Integer "tuya_get_protocol" dev))

(defn tuya-get-session-state [dev]
  (invoke Integer "tuya_get_session_state" dev))

(defn tuya-get-socket-state [dev]
  (invoke Integer "tuya_get_socket_state" dev))

(defn tuya-get-last-error [dev]
  (invoke Integer "tuya_get_last_error" dev))

;; -- Async mode -----------------------------------------------------

(defn tuya-set-async-mode [dev async]
  (invoke Void "tuya_set_async_mode" dev (boolean async)))

(defn tuya-is-socket-readable [dev]
  (ptr->bool (invoke Boolean "tuya_is_socket_readable" dev)))

(defn tuya-is-socket-writable [dev]
  (ptr->bool (invoke Boolean "tuya_is_socket_writable" dev)))

(defn tuya-set-session-ready [dev]
  (ptr->bool (invoke Boolean "tuya_set_session_ready" dev)))

;; -- Message building/decoding --------------------------------------

(defn tuya-build-message [dev buf cmd payload key]
  (invoke Integer "tuya_build_message" dev buf (int cmd) payload key))

(defn tuya-decode-message [dev buf size key]
  (invoke String "tuya_decode_message" dev buf (int size) key))

(defn tuya-generate-payload [dev cmd device-id datapoints]
  (invoke String "tuya_generate_payload" dev (int cmd) device-id datapoints))

;; -- Raw send/receive ------------------------------------------------

(defn tuya-send [dev buf size]
  (invoke Integer "tuya_send" dev buf (int size)))

(defn tuya-receive [dev buf maxsize minsize]
  (invoke Integer "tuya_receive" dev buf (int maxsize) (int minsize)))

;; -- device22 mode ---------------------------------------------------

(defn tuya-set-device22 [dev null-dps-json]
  (invoke Void "tuya_set_device22" dev null-dps-json))

(defn tuya-is-device22 [dev]
  (ptr->bool (invoke Boolean "tuya_is_device22" dev)))

;; -- High-level round-trip (auto-free strings) -----------------------

(defn tuya-set-value-bool [dev dp value]
  (consume-ptr (invoke Pointer "tuya_set_value_bool" dev (int dp) (boolean value))))

(defn tuya-set-value-int [dev dp value]
  (consume-ptr (invoke Pointer "tuya_set_value_int" dev (int dp) (int value))))

(defn tuya-set-value-string [dev dp value]
  (consume-ptr (invoke Pointer "tuya_set_value_string" dev (int dp) value)))

(defn tuya-set-value-float [dev dp value]
  (consume-ptr (invoke Pointer "tuya_set_value_float" dev (int dp) (double value))))

(defn tuya-turn-on
  ([dev] (tuya-turn-on dev 1))
  ([dev switch-dp]
   (consume-ptr (invoke Pointer "tuya_turn_on" dev (int switch-dp)))))

(defn tuya-turn-off
  ([dev] (tuya-turn-off dev 1))
  ([dev switch-dp]
   (consume-ptr (invoke Pointer "tuya_turn_off" dev (int switch-dp)))))

(defn tuya-status [dev]
  (consume-ptr (invoke Pointer "tuya_status" dev)))

(defn tuya-heartbeat [dev]
  (consume-ptr (invoke Pointer "tuya_heartbeat" dev)))

;; -- Type-aware set_value dispatcher ----------------------------------
;;
;;   (tuya-set-value dev dp :bool   true)
;;   (tuya-set-value dev dp :int    42)
;;   (tuya-set-value dev dp :string "hello")
;;   (tuya-set-value dev dp :float  3.14)
;;

(defn tuya-set-value
  "Set a device value by type keyword."
  [dev dp typ value]
  (case typ
    :bool   (tuya-set-value-bool   dev dp value)
    :int    (tuya-set-value-int    dev dp value)
    :string (tuya-set-value-string dev dp value)
    :float  (tuya-set-value-float  dev dp value)
    (throw (IllegalArgumentException. (str "Unknown type: " typ)))))

;; ==================================================================
;;  Constants
;; ==================================================================

;; Protocol versions
(def ^:const PROTO-V31 0)
(def ^:const PROTO-V33 1)
(def ^:const PROTO-V34 2)
(def ^:const PROTO-V35 3)

;; Session states
(def ^:const SESSION-INVALID      0)
(def ^:const SESSION-STARTING     1)
(def ^:const SESSION-FINALIZING   2)
(def ^:const SESSION-ESTABLISHED  3)

;; Socket states
(def ^:const SOCK-NO-SUCH-HOST   0)
(def ^:const SOCK-NO-SOCK-AVAIL  1)
(def ^:const SOCK-FAILED         2)
(def ^:const SOCK-DISCONNECTED   3)
(def ^:const SOCK-CONNECTING     4)
(def ^:const SOCK-CONNECTED      5)
(def ^:const SOCK-READY          6)
(def ^:const SOCK-RECEIVING      7)

;; Misc
(def ^:const DEFAULT-PORT           6668)
(def ^:const RECOMMENDED-BUFSIZE    1024)
(def ^:const DEFAULT-RETRY-LIMIT    5)
(def ^:const DEFAULT-RETRY-DELAY-MS 100)

;; Tuya command types (all 45)
(def ^:const CMD-UDP                     0)
(def ^:const CMD-AP-CONFIG               1)
(def ^:const CMD-ACTIVE                  2)
(def ^:const CMD-BIND                    3)
(def ^:const CMD-RENAME-GW               4)
(def ^:const CMD-RENAME-DEVICE           5)
(def ^:const CMD-UNBIND                  6)
(def ^:const CMD-CONTROL                 7)
(def ^:const CMD-STATUS                  8)
(def ^:const CMD-HEART-BEAT              9)
(def ^:const CMD-DP-QUERY               10)
(def ^:const CMD-QUERY-WIFI             11)
(def ^:const CMD-TOKEN-BIND             12)
(def ^:const CMD-CONTROL-NEW            13)
(def ^:const CMD-ENABLE-WIFI            14)
(def ^:const CMD-DP-QUERY-NEW           16)
(def ^:const CMD-SCENE-EXECUTE          17)
(def ^:const CMD-UPDATEDPS              18)
(def ^:const CMD-UDP-NEW                19)
(def ^:const CMD-AP-CONFIG-NEW          20)
(def ^:const CMD-GET-LOCAL-TIME         28)
(def ^:const CMD-WEATHER-OPEN           32)
(def ^:const CMD-WEATHER-DATA           33)
(def ^:const CMD-STATE-UPLOAD-SYN       34)
(def ^:const CMD-STATE-UPLOAD-SYN-RECV  35)
(def ^:const CMD-HEART-BEAT-STOP        37)
(def ^:const CMD-STREAM-TRANS           38)
(def ^:const CMD-GET-WIFI-STATUS        43)
(def ^:const CMD-WIFI-CONNECT-TEST      44)
(def ^:const CMD-GET-MAC                45)
(def ^:const CMD-GET-IR-STATUS          46)
(def ^:const CMD-IR-TX-RX-TEST          47)
(def ^:const CMD-LAN-GW-ACTIVE          240)
(def ^:const CMD-LAN-SUB-DEV-REQUEST    241)
(def ^:const CMD-LAN-DELETE-SUB-DEV     242)
(def ^:const CMD-LAN-REPORT-SUB-DEV     243)
(def ^:const CMD-LAN-SCENE              244)
(def ^:const CMD-LAN-PUBLISH-CLOUD-CONFIG  245)
(def ^:const CMD-LAN-PUBLISH-APP-CONFIG     246)
(def ^:const CMD-LAN-EXPORT-APP-CONFIG      247)
(def ^:const CMD-LAN-PUBLISH-SCENE-PANEL    248)
(def ^:const CMD-LAN-REMOVE-GW          249)
(def ^:const CMD-LAN-CHECK-GW-UPDATE    250)
(def ^:const CMD-LAN-GW-UPDATE          251)
(def ^:const CMD-LAN-SET-GW-CHANNEL     252)
